primitives/trie/src/cache.rs

```diff
@@ -0,0 +1,864 @@
+	pub fn new(cache_size: CacheSize) -> Self {
+		let (node_cache_size, value_cache) = match cache_size {
+			CacheSize::Maximum(max) => {
+				// The value cache element size isn't that huge, roughly around 50 bytes (mainly
+				// depending on the hash size). Thus, we only give it 5% of the overall cache size.
+				let value_cache_size_in_bytes = (max as f32 * 0.05) as usize;
+
+				(
+					CacheSize::Maximum(max - value_cache_size_in_bytes),
+					NoHashingLruCache::with_hasher(
+						value_cache_size_in_bytes / Self::value_cache_element_size(),
+						Default::default(),
+					),
+				)
+			},
+			CacheSize::Unlimited =>
+				(CacheSize::Unlimited, NoHashingLruCache::unbounded_with_hasher(Default::default())),
+		};
+
+		Self {
+			node_cache: Arc::new(RwLock::new(SharedNodeCache::new(node_cache_size))),
+			value_cache: Arc::new(RwLock::new(value_cache)),
+		}
+	}
+
+	/// Create a new [`LocalTrieCache`] instance from this shared cache.
+	pub fn local_cache(&self) -> LocalTrieCache<H> {
+		LocalTrieCache {
+			shared: self.clone(),
+			node_cache: Default::default(),
+			value_cache: Default::default(),
+			shared_node_cache_access: Default::default(),
+			shared_value_cache_access: Default::default(),
+		}
+	}
+
+	/// Returns the used memory size of this cache in bytes.
+	pub fn used_memory_size(&self) -> usize {
+		let value_cache_size = self.value_cache.read().len() * Self::value_cache_element_size();
+		let node_cache_size = self.node_cache.read().size_in_bytes;
+
+		node_cache_size + value_cache_size
+	}
+
+	/// Reset the node cache.
+	pub fn reset_node_cache(&self) {
+		self.node_cache.write().reset();
+	}
+
+	/// Reset the value cache.
+	pub fn reset_value_cache(&self) {
+		self.value_cache.write().clear();
+	}
+
+	/// Reset the entire cache.
+	pub fn reset(&self) {
+		self.reset_node_cache();
+		self.reset_value_cache();
+	}
+}
+
+/// The local trie cache.
+///
+/// This cache should be used per state instance created by the backend. One state instance is
+/// referring to the state of one block. It will cache all the accesses that are done to the state
+/// which could not be fullfilled by the [`SharedTrieCache`]. These locally cached items are merged
+/// back to the shared trie cache when this instance is dropped.
+///
+/// When using [`Self::as_trie_db_cache`] or [`Self::as_trie_db_mut_cache`], it will lock mutexes.
+/// So, it is important that these methods are not called multiple times, because they otherwise
+/// deadlock.
+pub struct LocalTrieCache<H: Hasher> {
+	/// The shared trie cache that created this instance.
+	shared: SharedTrieCache<H>,
+	/// The local cache for the trie nodes.
+	node_cache: Mutex<HashMap<H::Out, NodeOwned<H::Out>>>,
+	/// Keeps track of all the trie nodes accessed in the shared cache.
+	///
+	/// This will be used to ensure that these nodes are brought to the front of the lru when this
+	/// local instance is merged back to the shared cache.
+	shared_node_cache_access: Mutex<HashSet<H::Out>>,
```

Comment:
I am not exactly sure why we share this node access cache, is it for a more frequent access refresh? (would have seen it local and not shared at first glance)

Reply:
Yes, as the comment explains it is for refreshing the accessed values in the shared cache. We only lock the shared cache with read only to ensure that you can have concurrent access to it. Only when we merge stuff back, we lock the cache writeable. So, we need to keep track of the keys we have accessed to simulate the access later.